src/libopensc/card-jpki.c

```diff
@@ -0,0 +1,344 @@
+/*
+ * card-jpki.c: Support for JPKI(Japanese Individual Number Cards).
+ *
+ * Copyright (C) 2016, HAMANO Tsukasa <hamano@osstech.co.jp>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "internal.h"
+
+static struct sc_atr_table jpki_atrs[] = {
+	{"3b:e0:00:ff:81:31:fe:45:14", NULL, NULL,
+	 SC_CARD_TYPE_JPKI_BASE, 0, NULL},
+	{NULL, NULL, NULL, 0, 0, NULL}
+};
+
+static struct sc_card_operations jpki_ops;
+static struct sc_card_driver jpki_drv = {
+	"JPKI(Japanese Individual Number Cards)",
+	"jpki",
+	&jpki_ops,
+	NULL, 0, NULL
+};
+
+#define SELECT_MF 0
+#define SELECT_EFDIR 1
+
+struct jpki_private_data {
+	sc_file_t *mf;
+	sc_path_t aid;
+	int selected;
+};
+#define DRVDATA(card) ((struct jpki_private_data *) ((card)->drv_data))
+
+static int
+jpki_select_ap(struct sc_card *card)
+{
+	int rc;
+	sc_path_t path;
+
+	/* Select JPKI application */
+	sc_format_path("D392f000260100000001", &path);
+	path.type = SC_PATH_TYPE_DF_NAME;
+	rc = sc_select_file(card, &path, NULL);
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rc, "select JPKI AP failed");
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_match_card(struct sc_card *card)
+{
+	int i, rc;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	i = _sc_match_atr(card, jpki_atrs, &card->type);
+	if (i >= 0) {
+		return 1;
+	}
+	rc = jpki_select_ap(card);
+	if (rc == SC_SUCCESS) {
+		card->type = SC_CARD_TYPE_JPKI_BASE;
+		return 1;
+	}
+	return 0;
+}
+
+static int
+jpki_finish(sc_card_t * card)
+{
+	struct jpki_private_data *drvdata = DRVDATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (drvdata) {
+		free(drvdata);
+		card->drv_data = NULL;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_init(struct sc_card *card)
+{
+	struct jpki_private_data *drvdata;
+	sc_file_t *mf;
+	int flags;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	drvdata = malloc(sizeof (struct jpki_private_data));
+	if (!drvdata) {
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	memset(drvdata, 0, sizeof (struct jpki_private_data));
+	sc_format_path("D3 92 f0 00 26 01 00 00 00 01", &drvdata->aid);
```

Comment:
You're selecting this AID multiple times. Could you use something like a define or a global const for this?